0000-stream/0000-stream.md

```diff
@@ -0,0 +1,407 @@
+
+The receiving endpoint MUST close the connection with a `FlowControlError` if the sender violates the advertised limit.
+
+### 4.6. Closing Connections
+
+Either endpoint can close the connection using a `ConnectionClose` frame. Implementations MAY allow half-open connections (where one side has closed the connection and the other is still able to send).
+
+`ConnectionClose` frames are used to communicate both normal connection closes as well as errors.
+
+## 5. Packet and Frame Specification
+
+STREAM packets are encrypted and attached to ILP Prepare, Fulfill, and Reject packets as the `data` field.
+
+STREAM uses the [Octet Encoding Rules (OER)](http://www.oss.com/asn1/resources/books-whitepapers-pubs/Overview_of_OER.pdf) to encode packet fields.
+
+### 5.1. Encryption
+
+All STREAM packets are encrypted using [AES-256-GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) with a 12-byte Initialization Vector (IV) and a 16-Byte Authentication Tag.
+
+If subsequent versions support additional encryption algorithms, those details should be exchanged between the sender and receiver when they establish the shared secret. If a receiver attempts to decrypt an incoming packet but is unable to (perhaps because the sender is using an unsupported cipher), the receiver SHOULD reject the incoming transfer with an `F06: Unexpected Payment` error.
+
+#### 5.1.1. Encryption Envelope
+
+See the [ASN.1 definition](../asn1/Stream.asn) for the formal encryption envelope specification.
+
+| Field | Type | Description |
+|---|---|---|
+| Random IV  | 12-Byte UInt | Nonce used as input to the AES-GCM algorithm. Also ensures conditions are random. Endpoints MUST NOT encrypt two packets with the same nonce |
+| Authentication Tag | 16-Byte UInt | Authentication tag produced by AES-GCM encryption that ensures data integrity |
+| Ciphertext | 0-32739 Bytes | Encrypted data (see below for contents) |
+
+Note that the `Ciphertext` is NOT length-prefixed. The length can be inferred from the ILP packet because the whole `data` field is encoded with a length prefix. (This is done so that the entire `data` field is indistinguishable from random bytes.)
+
+#### 5.1.2. Encryption Pseudocode
+
+The encryption key used for every packet sent for a given connection is the HMAC-SHA256 digest of the shared secret and the string `"ilp_stream_encryption"`, encoded as UTF-8 or ASCII (the byte representation is the same with both encodings).
+
+```
+iv = random_bytes(12)
+encryption_key = hmac_sha256(shared_secret, "ilp_stream_encryption")
+{ ciphertext, auth_tag } = aes_256_gcm(encryption_key, iv, data)
+```
+
+#### 5.1.3. Maximum Number of Packets Per Connection
+
+Implementations MUST close the connection once either endpoint has sent 2^31 packets. According to [NIST](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf), it is unsafe to use AES-GCM for more than 2^32 packets using the same encryption key.
```

Comment:
is it 2^31 in order to give a bigger safety margin, or did you mean to write 2^32-1?

Reply:
Both sides are encrypting messages with the same key, so unless they keep track of how many packets they've gotten from the other side they should only send 2^31 messages each